technology-considerations/graphql-prototypes.md

```diff
@@ -0,0 +1,178 @@
+build artifacts.
+
+[CloudFormation]: https://aws.amazon.com/cloudformation/
+[tdr-terraform]: https://github.com/nationalarchives/tdr-prototype-terraform
+
+### Hosting
+
+Using AppSync would make the application extremely coupled to AWS
+Infrastructure, because so much of the code would be in Velocity templates and
+the rest would be implemented as AWS Lambdas.
+
+### Database connection
+
+TODO: Ian or Sam to add details of why it was hard to connect to the DB
+
+## Sangria approach
+
+As mentioned above, we prefer to use Scala as a backend language for consistency
+with our other projects. [Sangria] looks like a mature and feature-rich
+implementation of a GraphQL server in Scala, so we are using it to build a
+[GraphQL API prototype][sangria-prototype]. It uses [Slick] to map the Scala
+code to SQL queries.
+
+### Local development
+
+The development process is simple because it's just a single Scala application.
+We've wrapped the Sangria code in an [akka-http] server for development, which
+is quick to redeploy when modifying the code.
+
+We've added several API endpoints so far, and it's been fairly straightforward
+to build everything we need.
+
+### Deployment and hosting
+
+We've wrapped Sangria in a Lambda entry point for deployment, but we may switch
+to using akka-http in production as well. In that case, we would probably deploy
+a container to ECS, which we have experience of with the [TDR
+frontend][tdr-mvc].
+
+We currently deploy the application to Lambda by manually uploading a jar, but
+we don't anticipate it being difficult to automate. If we switch to ECS (or even
+a different hosting provider), we will deploy it in the same way as the TDR
+frontend.
+
+[Sangria]: https://sangria-graphql.org/
+[sangria-prototype]: https://github.com/nationalarchives/tdr-prototype-sangria
+[Slick]: http://slick.lightbend.com/
+[akka-http]: https://github.com/akka/akka-http
+[tdr-mvc]: https://github.com/nationalarchives/tdr-prototype-mvc
+
+## Conclusions
+
+Given how much easier it was to set up and develop a Sangria server than
+AppSync, we've decided to continue using Sangria to build a GraphQL server.
+
+## Client-side GraphQL experience
+
+This document is mostly about the choice of GraphQL server, but we've also
+learned a bit about GraphQL client libraries.
+
+Sangria has been easy to work with as a Scala GraphQL server library, but we
+haven't been able to find a good client library for Scala. The best option is
+[Drunk], but it appears to be a very thin wrapper around an HTTP client and
+the Circe JSON parser. You have to write graphQL queries as strings, with no
+type safety to protect you from syntax errors or inconsistencies between the
+code and the schema.
+
+For now, we've chosen to just use an HTTP client library directly. This gives
+us more control over the HTTP request, which is particularly useful when we need
+to do things like [sign requests][iam-signing].
+
```

Comment:
This [sbt plugin](https://github.com/muuki88/sbt-graphql) might 
allow generation of required classes from a schema.